src/adapter/src/catalog/storage.rs

```diff
@@ -199,28 +202,71 @@ impl Connection {
         bootstrap_args: &BootstrapArgs,
         deploy_generation: Option<u64>,
     ) -> Result<Connection, Error> {
+        let retry = Retry::default()
+            .clamp_backoff(Duration::from_secs(1))
+            .max_duration(Duration::from_secs(30))
+            .into_retry_stream();
+        let mut retry = pin::pin!(retry);
+        loop {
+            match Self::open_inner(stash, now.clone(), bootstrap_args, deploy_generation).await {
+                Ok(conn) => {
+                    return Ok(conn);
+                }
+                Err((given_stash, err)) => {
+                    stash = given_stash;
+                    let should_retry =
+                        matches!(&err, Error{ kind: ErrorKind::Stash(se)} if se.should_retry());
+                    if !should_retry || retry.next().await.is_none() {
+                        return Err(err);
+                    }
+                }
+            }
+        }
+    }
+
+    #[tracing::instrument(name = "storage::open_inner", level = "info", skip_all)]
+    async fn open_inner(
+        mut stash: Stash,
+        now: NowFn,
+        bootstrap_args: &BootstrapArgs,
+        deploy_generation: Option<u64>,
+    ) -> Result<Connection, (Stash, Error)> {
         // Initialize the Stash if it hasn't been already
-        let mut conn = if !stash.is_initialized().await? {
+        let mut conn = if !match stash.is_initialized().await {
+            Ok(is_init) => is_init,
+            Err(e) => {
+                return Err((stash, e.into()));
+            }
+        } {
```

Comment:
nit: all of these new `match` statements could use `.map_err(|e| (stash, e.into()))` instead

Reply:
That doesn't work because we move `stash` into the closure but then reference it further down

Reply:
Ahh frick you're so right! And open_inner has to take an owned `Stash` and not a `&mut Stash` because `Connection` needs to own it